blog-posts/typescript-ast-type-checker.md

```diff
@@ -0,0 +1,591 @@
+      ├── toLocaleString: { (): string; (locales?: string | string[], options?: DateTimeFormatOptions): string; }
+      ├── toLocaleDateString: { (): string; (locales?: string | string[], options?: DateTimeFormatOptions): string; }
+      ├── toLocaleTimeString: { (): string; (locales?: string | string[], options?: DateTimeFormatOptions): string; }
+      ├── valueOf: () => number
+      ├── getTime: () => number
+      ├── getFullYear: () => number
+      ├── getUTCFullYear: () => number
+      ├── getMonth: () => number
+      ├── getUTCMonth: () => number
+      ├── getDate: () => number
+      ├── getUTCDate: () => number
+      ├── getDay: () => number
+      ├── getUTCDay: () => number
+      ├── getHours: () => number
+      ├── getUTCHours: () => number
+      ├── getMinutes: () => number
+      ├── getUTCMinutes: () => number
+      ├── getSeconds: () => number
+      ├── getUTCSeconds: () => number
+      ├── getMilliseconds: () => number
+      ├── getUTCMilliseconds: () => number
+      ├── getTimezoneOffset: () => number
+      ├── setTime: (time: number) => number
+      ├── setMilliseconds: (ms: number) => number
+      ├── setUTCMilliseconds: (ms: number) => number
+      ├── setSeconds: (sec: number, ms?: number) => number
+      ├── setUTCSeconds: (sec: number, ms?: number) => number
+      ├── setMinutes: (min: number, sec?: number, ms?: number) => number
+      ├── setUTCMinutes: (min: number, sec?: number, ms?: number) => number
+      ├── setHours: (hours: number, min?: number, sec?: number, ms?: number) => number
+      ├── setUTCHours: (hours: number, min?: number, sec?: number, ms?: number) => number
+      ├── setDate: (date: number) => number
+      ├── setUTCDate: (date: number) => number
+      ├── setMonth: (month: number, date?: number) => number
+      ├── setUTCMonth: (month: number, date?: number) => number
+      ├── setFullYear: (year: number, month?: number, date?: number) => number
+      ├── setUTCFullYear: (year: number, month?: number, date?: number) => number
+      ├── toUTCString: () => string
+      ├── toISOString: () => string
+      ├── toJSON: (key?: any) => string
+      ├── getVarDate: () => VarDate
+      ├── __@toPrimitive@755: { (hint: "default"): string; (hint: "string"): string; (hint: "number"): number; (hint: string): string | number; }
+    👉├── value3: Date
+    👉├── collectedValue: ImportantValue
+    ```
+</details>
+
+That's the "standard library" issue described earlier. The `Date` and `string` types causes this drama and we need to stop our processing before entering those types.
+
+### Exclude the standard types
+
+TypeScript gives us plenty of tools to do that. Here is a helper method `isTypeLocal` I have build for our use cases.
+
+```typescript
+function isTypeLocal(symbol: ts.Symbol) {
+  const sourceFile = symbol?.valueDeclaration?.getSourceFile();
+  const hasSource = !!sourceFile;
+  const isStandardLibrary = hasSource && program.isSourceFileDefaultLibrary(sourceFile!)
```

Comment:
Is `isSourceFileDefaultLibrary` also a build in method? Wow, would be cool to have a small comment on how Typescript decides which library is default.

Reply:
I just see you explain it a bit in the next sentence 😄 A bit more detailed would be very interesting, like if a default library is everything the comes with Typescript.

Reply:
Hey I'm a consumer not a typescript contributor so I have no clue 😅 But a standard library is very well defined through this part in the tsconfig.json which should look familiar.

```
{
  "lib": ["es5", "es2015"]
}
```

I guess every file imported from there is marked with a flag to signal the standard library membership which can be queries through the inversion `file.hasNoDefaultLib`.

I added a link to the tsconfig lib documentation but I would like to focus on the API and chose not to dive deeper than necessary, although the entire TS processing is so damn interesting 🤩

Resources:
+ https://www.typescriptlang.org/tsconfig#lib 
+ https://github.com/microsoft/TypeScript/tree/main/lib
+ https://github.com/microsoft/TypeScript/blob/5afe42e14e61d7e4df5d75cc0022283711cb593a/src/compiler/program.ts#L1721-L1739